src/nfa/nfa.cc

```diff
@@ -193,21 +193,155 @@ void Nfa::add_trans(State stateFrom, Symbol symbolOnTransition, State stateTo) {
     auto transitionFromStateIter = transitionrelation[stateFrom].begin();
     for (; transitionFromStateIter != transitionrelation[stateFrom].end(); ++transitionFromStateIter) {
         if (transitionFromStateIter->symbol == symbolOnTransition) {
+            // Add transition with symbolOnTransition already used on transitions from stateFrom.
             transitionFromStateIter->states_to.insert(stateTo);
             return;
         } else if (transitionFromStateIter->symbol > symbolOnTransition) {
             break;
         }
     }
 
+    // Add transition to a new TransSymbolStates struct with symbolOnTransition yet unused on transitions from stateFrom.
     transitionrelation[stateFrom].insert(transitionFromStateIter, TransSymbolStates(symbolOnTransition, stateTo));
 }
 
+void Nfa::remove_trans(State src, Symbol symb, State tgt)
+{
+    if (!has_trans(src, symb, tgt))
+    {
+        throw std::invalid_argument(
+                "Transition [" + std::to_string(src) + ", " + std::to_string(symb) + ", " +
+                std::to_string(tgt) + "] does not exist.");
+    }
+
+    auto transitionFromStateIter{ transitionrelation[src].begin() };
+    while (transitionFromStateIter != transitionrelation[src].end())
+    {
+        if (transitionFromStateIter->symbol == symb)
+        {
+            transitionFromStateIter->states_to.remove(tgt);
+            if (transitionFromStateIter->states_to.empty())
+            {
+                transitionrelation[src].remove(*transitionFromStateIter);
+            }
+
+            return;
+        }
+
+        ++transitionFromStateIter;
+    }
+}
+
 State Nfa::add_new_state() {
     transitionrelation.emplace_back();
     return transitionrelation.size() - 1;
 }
 
+void Nfa::remove_epsilon(const Symbol epsilon)
+{
+    *this = Mata::Nfa::remove_epsilon(*this, epsilon);
+}
+
+StateSet Nfa::get_reachable_states() const
+{
+    StateBoolArray reachable_bool_array{ compute_reachability() };
+
+    StateSet reachable_states{};
+    for (State original_state{ 0 }; original_state < get_num_of_states(); ++original_state)
+    {
+        if (reachable_bool_array[original_state])
+        {
+            reachable_states.push_back(original_state);
+        }
+    }
+
+    return reachable_states;
+}
+
+StateSet Nfa::get_terminating_states() const
+{
+    Nfa reversed{ revert(*this) };
+    return reversed.get_reachable_states();
+}
+
+void Nfa::trim()
+{
+    StateSet original_useful_states{ get_useful_states() };
+
+    StateMap<State> original_to_new_states_map{ original_useful_states.size() };
+    size_t new_state_num{ 0 };
+    for (const State original_state: original_useful_states)
+    {
+        original_to_new_states_map.insert(std::make_pair(original_state, new_state_num));
+        ++new_state_num;
+    }
+
+    Nfa trimmed_aut{ initialize_trimmed_aut(original_to_new_states_map) };
+
+    add_trimmed_transitions(original_to_new_states_map, trimmed_aut);
+
+    *this = trimmed_aut;
```

Comment:
Same applies as above. Seems dangerous.

Reply:
I will investigate.

Reply:
We agreed that the use of `*this = ` in this context is safe. For reasons for this decision, see the other discussion on this topic above.